src/selectors/getSortPreference.ts

```diff
@@ -1,31 +1,62 @@
+import SortDirection from '../@types/SortDirection'
 import SortPreference from '../@types/SortPreference'
 import State from '../@types/State'
 import ThoughtId from '../@types/ThoughtId'
 import findDescendant from '../selectors/findDescendant'
-import parseSortDirection from '../util/parseSortDirection'
+import getThoughtById from '../selectors/getThoughtById'
 import { getAllChildrenAsThoughts } from './getChildren'
 import getGlobalSortPreference from './getGlobalSortPreference'
 
 /**
  * Get sort direction if given sort type is not 'None'.
  */
-const getSortDirection = (sortType: string, state: State, id: ThoughtId) => {
+const getSortDirection = (sortType: string, state: State, sortId: ThoughtId): SortDirection | null => {
   if (sortType === 'None') return null
-  const sortTypeId = findDescendant(state, id, ['=sort', sortType])
-  const childrenSortDirection = sortTypeId ? getAllChildrenAsThoughts(state, sortTypeId) : []
-  return childrenSortDirection.length > 0 ? parseSortDirection(childrenSortDirection[0].value) : 'Asc'
+
+  // Start from sortId and look for the sortType
+  const sortTypeId = findDescendant(state, sortId, [sortType])
+  if (!sortTypeId) return 'Asc'
+
+  const childrenSortDirection = getAllChildrenAsThoughts(state, sortTypeId)
+  const directionChild = childrenSortDirection.find(child => child.value === 'Desc' || child.value === 'Asc')
+
+  return directionChild ? (directionChild.value as SortDirection) : 'Asc'
+}
+
+/** Find the parent ID of a given thought ID. */
+const findParentId = (state: State, thoughtId: ThoughtId): ThoughtId | null => {
+  const thought = getThoughtById(state, thoughtId)
+  return thought ? thought.parentId : null
 }
 
 /** Get the sort setting from the given context meta or, if not provided, the global sort. */
 const getSortPreference = (state: State, id: ThoughtId): SortPreference => {
-  const sortId = findDescendant(state, id, ['=sort'])
-  const childrenSort = sortId ? getAllChildrenAsThoughts(state, sortId) : []
-  return childrenSort.length > 0
-    ? {
-        type: childrenSort[0].value,
-        direction: getSortDirection(childrenSort[0].value, state, id),
-      }
-    : getGlobalSortPreference(state)
+  let sortId = findDescendant(state, id, ['=sort'])
+  let childrenSort = sortId ? getAllChildrenAsThoughts(state, sortId) : []
+
+  // If no direct sort is found, iteratively check parent context
+  let currentId: ThoughtId | null = id
+  while ((!sortId || childrenSort.length === 0) && currentId && currentId !== '__ROOT__') {
```

Comment:
Can you explain the the while loop? It looks like it checks every ancestor, but sort is only supposed to be determined by the parent's meta attributes.

Reply:
The while loop was initially designed to traverse up the ancestor chain to check for a =sort meta attribute. However, you're correct that sorting should only depend on the immediate parent's =sort meta attribute.

In the current implementation:

1. The loop checks every ancestor context until it either finds a =sort meta attribute with children or reaches the root.
2. This logic attempts to handle cases where no direct =sort is found for the current context or its immediate parent.

To align with the intended behavior - restricting sort determination to the immediate parent's =sort - the logic can be simplified to:

- Check the current context for =sort.
- If not found or empty, check the immediate parent's =sort meta.

The while loop is no longer necessary for this behavior. I’ve updated the logic accordingly in the revised version. Let me know if there’s any edge case you think might still require the loop.

Reply:
Yes, that's closer to the expected behavior.

Maybe I'm missing something, but what was wrong with the original `getSortPreference`? It has been working fine as far as I know.

Reply:
Previously, getCursorSortDirection used cursorId directly, which didn't yield the correct sort direction. By updating it to use the parent of cursorId, the function now returns the correct sort direction. The new version works as intended.

Reply:
Since getCursorSortDirection in Sort.tsx has been updated to use parentId, the previous adjustments to getSortPreference are unnecessary.